cpp/src/arrow/compute/exec/expression.h

```diff
@@ -129,6 +129,20 @@ class ARROW_EXPORT Expression {
   explicit Expression(Datum literal);
   explicit Expression(Parameter parameter);
 
+  /*
+    Grammar:
+    Expr -> FieldRef | Literal | Call
+
+    FieldRef -> Field | Field FieldRef
+    Field -> . Name | [ Number ]
+
+    Literal -> $ TypeName : Value
```

Comment:
This is missing rules for escaping right?  I think those need to be part of the grammar.